---
title: "Introduction to PhyloVision and Hotspot"
package: "`r BiocStyle::pkg_ver('BiocStyle')`"
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Introduction to PhyloVision and Hotspot}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Introduction
As of version 3.0.0, VISION now supports the analysis of scRNA-seq data with respect to a user-defined cell lineage, or phylogeny. In this case, VISION uses the relationships between cells as specified by the phylogeny, as opposed to some user-defined latent space, to conduct the autocorrelation analysis. In this vignette, we demonstrate how a user might use this variant of the VISION pipeline - which we term `PhyloVision` - to analyze a recently published dataset that simultaneously profiled the single-cell transcriptomes and lineages of mouse embryogenesis (Chan et al, Nature 2019).

This tutorial begins by showing users how to create a `PhyloVision` object and performing analysis on it. Then, we demonstrate how this object can be passed into the Hotspot pipeline that will identify de-novo transcriptional gene sets that are autocorrelated on the tree (this is described more below and in the [Spatial Hotspot vignette](spatialHotspot.html). Finally, we show users how to launch a report for viewing.

# Preliminaries

If you have yet to install VISION, we recommend installing the package from Github to install this package. Full source code can be found at the VISION Github repository, available [here](http://www.github.com/YosefLab/VISION).

```r{eval=FALSE}
require(devtools)
install_github("YosefLab/VISION")
```

Once VISION and R are installed, you may load in VISION using `library(VISION)`.

To enable the Hotspot analysis below, install it directly from the git repository using the following command:

```
pip install git+https://github.com/yoseflab/Hotspot.git
```
If you are having trouble installing Hotspot, please refer to the documentation website [here](https://yoseflab.github.io/Hotspot/).

# Running PhyloVision

First, we need to load VISION, reticulate and ape.
```{r setup, eval=F}
knitr::opts_chunk$set(echo = TRUE)
library(VISION)
library(reticulate)
library(ape)
```

## Creating a PhyloVision Object with a Tree

Vision objects now support dendrograms for visualization and analysis. To create the PhyloVision object, you need all of the same data, an expression matrix, and metadata and/or signatures, but you also include a tree object from the Ape package.

```{r create, eval=F}
# Read the expression and meta data
file_path <- "data/embryogenesis" # Replace the path

expr <- read.table(paste(file_path, "expr_filtered.tsv.gz", sep="/"), check.names = FALSE, sep = "\t")

meta <- read.table(paste(file_path, "meta.tsv", sep="/"), check.names = FALSE, sep = "\t", row.names = 1, header=TRUE)
meta$Cell_Type <- as.factor(meta$Cell_Type)
# Signature file
sig <- paste("data", "h.all.v5.2.symbols.gmt", sep="/")

# Read the tree
tree <- read.tree(paste(file_path, "embryo_tree.newick", sep="/"))

# Collapse one mutations with ape
tree <- collapse.singles(tree)

# Construct the Vision object
vis <- PhyloVision(tree=tree, data=expr, signatures=sig, meta=meta, num_neighbors=30, projection_genes= rownames(expr))
```
**Tree**

The `ape` phylogeny object. Singleton edges should be collapsed prior to use. Ensure that leaf names correspond to same names used in the expression matrix. 

**Expression Data**

The provided expression data should be library-normalized. The expression data should not be log-transformed prior to loading into VISION. For more information on the input to VISION, please refer to the central [VISION vignette](VISION_vignette.html).

**Signatures**

Signatures can be provided as a list of paths to signature files (\*.gmt) or Signature objects.

See [the signature vignette](Signatures.html) for more information on finding or creating gene Signatures.

**Meta Data**

An R data.frame with cell-level meta-data.  This could be confounding variables (e.g. percentage of mitochondrial RNA, number of genes detected) or experimental covariates (e.g. genotype, donor, batch).

This input is optional if Signatures are provided.

**Other Options**

Other options and inputs can be provided to customize how VISION runs.

## Running PhyloVision analysis

Next, we can perform the normal Vision analysis using the tree as the latent space.

``` {r analyze, eval=F}
vis <- phyloAnalyze(vis)
```
## Running Hotspot analysis

As mentioned above, PhyloVision can take advantage of the new Hotspot functionality in VISION. Briefly, Hotspot is a tool for inferring modules of genes that are significantly autocorrelated with one another and a particular latent space (e.g., the first 30 principal components of a gene expression matrix). When combined with PhyloVision, the Hotspot functionality will use the user-defined phylogeny as the latent space.

We can invoke the Hotspot analysis with the function `runHotspot` and setting `tree=True`. Upon doing so, this will identify modules of genes and add these as Signatures to the VISION object for autocorrelation evaluation. Moreover, to add interpretability to this analysis, VISION will also compute the enrichment between all the user-defined Signatures and each Hotspot module. This information will be accessible on the web-based report by selecting the "Hotspot" mode in the top-right of the web-page.

The full Hotspot API is exposed for analysis as well. For more on the Hotspot API see [here](https://yoseflab.github.io/Hotspot/index.html). To note, while it is typically not recommended to use `model="normal"` and `logdata=FALSE` in Hotpsot, we elect to do so because the data here has been previously log-normalized. In more typical single-cell anlayses where the user is working with a library-normalized count matrix, we suggest setting `model="danb"` and `logdata=TRUE`.

```{r hotspot, eval=F}
vis <- runHotspot(vis, model="normal", tree=TRUE, min_gene_threshold=50, n_neighbors = 30, number_top_genes = nrow(expr), logdata=FALSE)
```

We have additionally exposed the full Hotspot pipeline should a user want to iterate with different parameters:
```{r full_h, eval=FALSE}
# Init Hotspot
hs <- hsInit(vis, model = "normal", tree=TRUE, logdata=FALSE)
# Init Hotspot KNN
hs <- hsCreateKnnGraph(hs, vis, n_neighbors=30)
# perform Hotspot analysis and store results in R
hs_genes <- hsComputeAutoCorrelations(hs, number_top_genes=12438)
# Compute localcorr
hs <- hsComputeLocalCorrelations(hs, hs_genes)
# Calculate Hotspot Module Scores for informative genes
hs <- hsCalculateModuleScores(hs)
# Cluster Hotspot modules and perform Vision based analysis on HS Modules and 
vis <- analyzeHotspotObjectVision(vis, hs, tree=TRUE)
```

We can also access further Hotspot functionality using Reticulate and Python.
```{r inspectModules, eval=F}
hs <- loadHotspotObject(bytes=vis@Hotspot)
library(reticulate)
use_python('/usr/bin/python3')
```
```{python modulesPlot, eval=F}
import matplotlib.pyplot as plt
import hotspot
hs.plot_local_correlations()
plt.show()
```
Note: the heatmap can also be visualized like so:
```{r heatmapR, eval=F}
hs <- loadHotspotObject(bytes=vis@Hotspot)
library(paletteer)
draw_hotspot_heatmap(hs)
```

## Viewing Results

Finally, we can launch the Vision browser. If you are local, the following will work:
```{r eval=FALSE}
viewResults(vis)
```

Else, if you would like to launch from a server, you can use the following settings:
```{r, eval=FALSE}
viewResults(vis, host='0.0.0.0', port=8100, browser=F)
```
Now, the report will be visible from port `8100` from your server (e.g., http://server-name:8100/Results.html)
