<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Signature Autocorrelation • VISION</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><link href="../../extra.css" rel="stylesheet">
<meta property="og:title" content="Signature Autocorrelation">
<meta property="og:description" content="VISION">
<meta property="og:image" content="https://yoseflab.github.io/VISION/logo.svg">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">VISION</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">3.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../../articles/VISION-vignette.html">Getting Started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/Signatures.html">Gene Signatures</a>
    </li>
    <li>
      <a href="../../articles/web_only/SignatureAutocorrelation.html">Signature Autocorrelation</a>
    </li>
    <li>
      <a href="../../articles/micropooling.html">Micropooling</a>
    </li>
    <li>
      <a href="../../articles/trajectories.html">Incorporating a Trajectory Model</a>
    </li>
    <li>
      <a href="../../articles/phyloVision.html">PhyloVision and Hotspot</a>
    </li>
    <li>
      <a href="../../articles/spatialHotspot.html">Spatial Data and Hotspot</a>
    </li>
    <li>
      <a href="../../articles/web_only/Seurat.html">How To: Working with Seurat</a>
    </li>
    <li>
      <a href="../../articles/web_only/10xData.html">How To: 10x CellRanger Outputs</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
<li>
  <a href="../../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/YosefLab/VISION/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="SignatureAutocorrelation_files/header-attrs-2.8/header-attrs.js"></script><script src="SignatureAutocorrelation_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Signature Autocorrelation</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/YosefLab/VISION/blob/master/vignettes/web_only/SignatureAutocorrelation.Rmd"><code>vignettes/web_only/SignatureAutocorrelation.Rmd</code></a></small>
      <div class="hidden name"><code>SignatureAutocorrelation.Rmd</code></div>

    </div>

    
    
<style>
.center-img img {
    display: block;
    margin: 50px auto 0 auto;
}
.center-img p.caption {
    text-align: center;
    margin: 20px 80px 40px 80px;
}
</style>
<div id="what-is-signature-autocorrelation" class="section level1">
<h1 class="hasAnchor">
<a href="#what-is-signature-autocorrelation" class="anchor"></a>What is signature autocorrelation?</h1>
<p>When analyzing single-cell data using signatures, one goal would be to identify signatures whose values exhibit interesting patterns of variation in the dataset.</p>
<p>A labor-intensive way to do this would be to plot the expression scores of every signature on top of a visualization of the data (e.g. with tSNE or UMAP) and identify signatures whose scores vary in a coordinated manner. The idea being that signatures that exhibit this property can be used to potentially describe features of the latent space.</p>
<p>VISION both automates this procedure and makes the process more quantitative. A measure of global autocorrelation is evaluated for each signature in the latent space model. This allows for a natural ranking of signatures as well as testing for significance.</p>
<p>To better understand this statistic, which varies from 0 (no autocorrelation to 1 (perfect autocorrelation), the graphic below shows a projection of single cells overlaid with signatures of 6 different autocorrelation values (labeled as C’).</p>
<div class="center-img">
<div class="figure">
<img src="cvalue_demo.png" alt=""><p class="caption">Higher values of C’ indicate higher signature autocorrelation. Here the expression of CD4 T cells is visualized using tSNE and colored according to signature score.</p>
</div>
</div>
</div>
<div id="computing-autocorrelation" class="section level1">
<h1 class="hasAnchor">
<a href="#computing-autocorrelation" class="anchor"></a>Computing Autocorrelation</h1>
<p>Signature autocorrelation captures the degree to which a cell’s signature score is similar to that of its ‘nearest neighbors’ in a latent model of gene expression. Or put another way, “Do cells with similar global expression also tend to have similar values for the gene signature being examined?” Signatures whose variation is largely composed of random technical noise would not behave in this way and cell-specific biases are controlled for in the calculation of signature scores (described <a href="Signatures.html">here</a>).</p>
<p>As a measure of autocorrelation, we use the <a href="https://en.wikipedia.org/wiki/Geary%27s_C">Geary’s C</a> statistic. To compute this statistic, two pieces of information are needed:</p>
<ul>
<li>‘weights’ between measurements indicating how similar (e.g., nearby in the latent model) cells are to one another</li>
<li>values to be compared against these weights</li>
</ul>
<p>The values consist of the <a href="Signatures.html">signature scores</a>. To calculate the weights, first the nearest <span class="math inline">\(k\)</span> neighbors are determined for each cell in the latent model. If a latent space is provided as the latent model (or if it is computed using PCA), then nearest neighbors are determined using euclidean distance. If a trajectory model is instead provided, neighbors are determined using the distance between cells along the trajectory. Weights are then computed by applying a Gaussian kernel to the distances between nearest neighbors such that higher weights are assigned to similar cells and the weight between distant cells is reduced to zero.</p>
<p>Finally, the autocorrelation statistic, C’ is computed as:</p>
<p><span class="math display">\[
C' = 1-\frac{
    (N-1)\sum_i \sum_j w_{ij}(x_i-x_j)^2
}{
    2W\sum_i(x_i-\bar{x})^2
}
\]</span></p>
<ul>
<li>
<span class="math inline">\(x_{i}\)</span> represents the signature score of cell <span class="math inline">\(i\)</span>
</li>
<li>
<span class="math inline">\(w_{ij}\)</span> is the weight between cells <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span>
</li>
<li>
<span class="math inline">\(W\)</span> is the sum of all weights</li>
<li>
<span class="math inline">\(N\)</span> is the number of cells</li>
<li>
<em>Note</em>: This is actually 1 minus the Geary’s C (which is why we label it as <span class="math inline">\(C'\)</span> instead of just <span class="math inline">\(C\)</span>). This alteration was made so that no autocorrelation occurs at a value of zero and maximal autocorrelation occurs at a value of one - which we found to be more intuitive due to its connection with other measures of correlation.</li>
</ul>
<p>A permutation test using randomly-generated signatures (of similar size and positive-to-negative ratio) is used to assess the significance of these values. However, due to the large number of cells involved in modern single-cell experiments, it is fairly easy to reject the null hypothesis and so the authors also recommend ranking and filtering using the C’ values and not only the FDR rates (similar to how differential expression tests may be further filtered by the estimated log fold-change).</p>
<p>More details are <a href="https://www.biorxiv.org/content/early/2018/09/28/403055">available in our Biorxiv preprint.</a></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Matt Jones, David Detomaso, Tal Ashuach, Yanay Rosen.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
