<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>GET: /SessionInfo • VISION</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script>



  <link href="extra.css" rel="stylesheet">
  

<meta property="og:title" content="GET: /SessionInfo" />
<meta property="og:image" content="https://yoseflab.github.io/VISION/logo.svg" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-title-body">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">VISION</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">3.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="articles/VISION-vignette.html">Getting Started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="articles/Signatures.html">Gene Signatures</a>
    </li>
    <li>
      <a href="articles/web_only/SignatureAutocorrelation.html">Signature Autocorrelation</a>
    </li>
    <li>
      <a href="articles/micropooling.html">Micropooling</a>
    </li>
    <li>
      <a href="articles/trajectories.html">Incorporating a Trajectory Model</a>
    </li>
    <li>
      <a href="articles/phyloVision.html">PhyloVision and Hotspot</a>
    </li>
    <li>
      <a href="articles/spatialHotspot.html">Spatial Data and Hotspot</a>
    </li>
    <li>
      <a href="articles/web_only/Seurat.html">How To: Working with Seurat</a>
    </li>
    <li>
      <a href="articles/web_only/10xData.html">How To: 10x CellRanger Outputs</a>
    </li>
  </ul>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/YosefLab/VISION/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>GET: /SessionInfo</h1>
    </div>


<div id="get-sessioninfo" class="section level1">

<p>Returns high-level configuration info on this VISION session. Mostly used to enable/disable features in the visualizer.</p>
<ul>
<li>‘name’: A title for this session (shown in the top title bar)</li>
<li>‘meta_sigs’: Names of meta-data signatures. Sent here as a convenience as numeric meta-data signatures and gene signatures get combined in the display and often lookups are needed to distinguish.</li>
<li>‘ncells’: Total number of cells</li>
<li>‘pooled’: Whether or not there is micro-pooling</li>
<li>‘has_sigs’: Whether or not there are gene signatures</li>
<li>‘has_proteins’: Whether or not there are CITE-seq proteins</li>
<li>‘has_lca’: Whether or not LC Annotator analysis is available</li>
<li>‘has_tree’: Whether or not Trajectories are available</li>
</ul>
<p>Response structure:</p>
<pre><code>    {
        'name': 'Session Name',
        'meta_sigs': [list of string],
        'ncells': int,
        'pooled': bool,
        'has_sigs': bool,
        'has_proteins': bool,
        'has_lca': bool,
        'has_tree': bool,
    }</code></pre>
</div>
<div id="get-signaturescores" class="section level1">
<h1 class="hasAnchor">
<a href="#get-signaturescores" class="anchor"></a>GET: /Signature/Scores/<sig_name></sig_name>
</h1>
<p>Return the per-cell signature scores for signature: <code>&lt;sig_name&gt;</code> Data comes from <a href="mailto:obj@SigScores" class="email">obj@SigScores</a> matrix</p>
<p>Response structure:</p>
<pre><code>    {
        'cells': [list of str,  cell ids...],
        'values': [list of Number, per cell signature scores],
    }</code></pre>
</div>
<div id="get-signaturemeta" class="section level1">
<h1 class="hasAnchor">
<a href="#get-signaturemeta" class="anchor"></a>GET: /Signature/Meta/<sig_name></sig_name>
</h1>
<p>Return the per-cell meta-data values for variable: <code>&lt;sig_name&gt;</code> Data comes from <a href="mailto:obj@metaData" class="email">obj@metaData</a> dataframe</p>
<p>Response structure:</p>
<pre><code>    {
        'cells': [list of str,  cell ids...],
        'values': [list of Number or string, per-cell meta-data values],
    }</code></pre>
</div>
<div id="get-signatureinfo" class="section level1">
<h1 class="hasAnchor">
<a href="#get-signatureinfo" class="anchor"></a>GET: /Signature/Info/<sig_name></sig_name>
</h1>
<p>Returns information associated with a signature Data comes from <a href="mailto:object@sigData" class="email">object@sigData</a></p>
<p>Response structure:</p>
<pre><code>    {
        'sigDict': { # signature-gene weights
            'GeneName1': 1, 
            'GeneName2': -1,
            ...
        },
        'name': 'Signature Name',
        'source': 'Signature source', # typically filename where signature was loaded from
        'metaData': 'Signature description...', # second column in .gmt file
        'geneImportance': { # signature-gene importance values
            'GeneName1': .1,
            'GeneName2': .01, 
            ...
        }
    }</code></pre>
</div>
<div id="get-signatureexpression" class="section level1">
<h1 class="hasAnchor">
<a href="#get-signatureexpression" class="anchor"></a>GET: /Signature/Expression/<sig_name>/<cluster_var></cluster_var></sig_name>
</h1>
<p>Returns expression values for <code>&lt;sig_name&gt;</code> grouped by <code>&lt;cluster_var&gt;</code>. This is used to generate the gene x cluster heatmap in the bottom left Values are computed on the fly at the server</p>
<p>Response structure</p>
<pre><code>{
    'data': list of list of Number  (2d matrix, outer index is row index)
    'sample_labels': column labels for the matrix in 'data'
    'gene_labels': row labels for the matrix in 'data'
}</code></pre>
</div>
<div id="get-proteinsprotein_namevalues" class="section level1">
<h1 class="hasAnchor">
<a href="#get-proteinsprotein_namevalues" class="anchor"></a>GET: /Proteins/&lt;protein_name&gt;/Values</h1>
<p>Returns the per-cell protein expression values for <code>&lt;protein_name&gt;</code> Data comes from <a href="mailto:obj@proteinData" class="email">obj@proteinData</a> matrix</p>
<p>Response structure:</p>
<pre><code>    {
        'cells': [list of str,  cell ids...],
        'values': [list of Number, per cell protein expression],
    }</code></pre>
</div>
<div id="get-filtergroupsigclustersnormal" class="section level1">
<h1 class="hasAnchor">
<a href="#get-filtergroupsigclustersnormal" class="anchor"></a>GET: /FilterGroup/SigClusters/Normal</h1>
<p>For each signature, describes the assignment of signatures to signature-clusters This is used to group signatures in the upper-left section</p>
<p>Response structure:</p>
<pre><code>{
    "Signature Name" (str): &lt;cluster_number&gt; (int),
    ...
}</code></pre>
</div>
<div id="get-filtergroupsigclustersproteins" class="section level1">
<h1 class="hasAnchor">
<a href="#get-filtergroupsigclustersproteins" class="anchor"></a>GET: /FilterGroup/SigClusters/Proteins</h1>
<p>For each protein, describes the assignment of proteins to protein-clusters Proteins aren’t actually clustered in the interface. Every protein is assigned to cluster #1</p>
<p>Response structure:</p>
<pre><code>{
    "Protein Name" (str): 1,
    ...
}</code></pre>
</div>
<div id="get-filtergroupsigclustersmeta" class="section level1">
<h1 class="hasAnchor">
<a href="#get-filtergroupsigclustersmeta" class="anchor"></a>GET: /FilterGroup/SigClusters/Meta</h1>
<p>For each meta-data variable, describes the assignment of each variable to meta-data clusters This is used when micropools are created to group factor variables with their expanded percent-value-per-micropool representations</p>
<p>Response structure:</p>
<pre><code>{
    "Metadata variable Name" (str): &lt;cluster_number&gt; (int),
    ...
}</code></pre>
</div>
<div id="get-projectionscoordinates" class="section level1">
<h1 class="hasAnchor">
<a href="#get-projectionscoordinates" class="anchor"></a>GET: /Projections/<proj_name>/coordinates/<proj_col></proj_col></proj_name>
</h1>
<p>Returns the coordinates for projection <code>&lt;proj_name&gt;</code> in column <code>&lt;proj_col&gt;</code> Since the scatter plot can show more than just projects, this has been expanded to also return coordinates from the latent space or other meta-data variables. <code>&lt;proj_name&gt;</code> can come from names in <a href="mailto:object@Projections" class="email">object@Projections</a>, or can be “Meta Data” to refer to <a href="mailto:object@metaData" class="email">object@metaData</a> or “Latent Space” to refer to <a href="mailto:object@LatentSpace" class="email">object@LatentSpace</a>. In either case <code>&lt;proj_name&gt;</code> should refer to an associated column.</p>
<p>Response structure:</p>
<p>List of list of <code>[coordinate, 'cell id']</code></p>
<pre><code>[
    [coordinate (float): 'cell id' (str)],
    [coordinate (float): 'cell id' (str)],
    [coordinate (float): 'cell id' (str)],
    ...
[</code></pre>
</div>
<div id="get-projectionslist" class="section level1">
<h1 class="hasAnchor">
<a href="#get-projectionslist" class="anchor"></a>GET: /Projections/list</h1>
<p>Returns information on projections and other variables that can be plotted in the main scatter-plot panel. Used to create the dropdown menus.</p>
<p>Response structure:</p>
<p>Dictionary where keys are the name of projections/meta data/latent space and the values are a list of the associated columns</p>
<pre><code>{
    'projection_name_1': list of column names in the projection,
    'projection_name_2': list of column names in the projection,
    'Meta Data': list of numeric meta-data variables,
    'Latent Space': list of column names in the latent space
}</code></pre>
</div>
<div id="get-pearsoncorrnormal" class="section level1">
<h1 class="hasAnchor">
<a href="#get-pearsoncorrnormal" class="anchor"></a>GET: /PearsonCorr/Normal</h1>
<p>Gets information for the LC Annotator - the correlation of each signature with each latent component</p>
<p>Response structure:</p>
<pre><code>{
    "zscores": list of list of correlation coefficients (N signatures x N components),
    "pvals": list of list of p-values (N signatures x N components),
    "proj_labels": list of latent space column names in the same order as columns of 'zscores'/'pvals',
    "sig_labels": list of signature names in the same order as columns of 'zscores'/'pvals',
}</code></pre>
</div>
<div id="get-pearsoncorrproteins" class="section level1">
<h1 class="hasAnchor">
<a href="#get-pearsoncorrproteins" class="anchor"></a>GET: /PearsonCorr/Proteins</h1>
<p>Gets information for the LC Annotator - the correlation of each protein with each latent component</p>
<p>Response structure:</p>
<pre><code>{
    "zscores": list of list of correlation coefficients (N proteins x N components),
    "pvals": list of list of p-values (N proteins x N components),
    "proj_labels": list of latent space column names in the same order as columns of 'zscores'/'pvals',
    "sig_labels": list of proteins names in the same order as columns of 'zscores'/'pvals',
}</code></pre>
</div>
<div id="get-pearsoncorrmeta" class="section level1">
<h1 class="hasAnchor">
<a href="#get-pearsoncorrmeta" class="anchor"></a>GET: /PearsonCorr/Meta</h1>
<p>Gets information for the LC Annotator - the correlation of each numeric meta-data variable with each latent component</p>
<p>Response structure:</p>
<pre><code>{
    "zscores": list of list of correlation coefficients (N variables x N components),
    "pvals": list of list of p-values (N variables x N components),
    "proj_labels": list of latent space column names in the same order as columns of 'zscores'/'pvals',
    "sig_labels": list of variables names in the same order as columns of 'zscores'/'pvals',
}</code></pre>
</div>
<div id="post-de" class="section level1">
<h1 class="hasAnchor">
<a href="#post-de" class="anchor"></a>POST: /DE</h1>
<p>Requests a differential expression analysis from the server.</p>
<p>Determines the set of cells used for the numerator (i.e. _num or _n) and for the denominator (i.e. _denom or _d)</p>
<p>Examples:</p>
<ul>
<li>To select cells in cluster 5 for the numerator set <code>'type_n': 'meta'</code>, <code>'subtype_n': 'cluster'</code>, <code>'group_num': 'Cluster 5'</code>
</li>
<li>To select cells in a saved manual selection for the numerator set <code>'type_n': 'saved_selection'</code>, <code>'group_num': 'name_of_selection'</code>
</li>
<li>To select a specific set of cell ids for the numerator set <code>'type_n': 'current'</code>, <code>'group_num': list of cell ids</code>
</li>
</ul>
<p>Posted data structure:</p>
<pre><code>{
    'type_n': either 'current', 'meta', or 'saved_selection'
    'type_d':  either 'remainder', 'meta', or 'saved_selection'
    'subtype_n': which meta-data variable to group cells by if type_n is 'meta'
    'subtype_d': which meta-data variable to group cells by if type_d is 'meta'
    'group_num': list of cell ids (if type_n is 'current') OR name of saved selection (if type_n is 'saved_selection')
                 OR value of categorical meta-data variable (if type_n is 'meta')
    'group_denom': same behavior as 'group_num'
    'min_cells': (int) pre-filter genes expressed in less than this many cells
    'subsample_groups': (bool), whether or not to subsample each comparison group first
    'subsample_N': (int) size of the groups after subsampling.  Ignored if 'subsample_groups' is false
}</code></pre>
<p>Response structure:</p>
<p>DE Results table with {‘column name’ -&gt; list of values}</p>
<pre><code>{
    'Feature': list of feature (gene or protein) names
    'Type': list of either 'Gene' or 'Protein' 
    'logFC': list of log-fold-change values
    'stat': list of AUC values
    'pval': list of FDR-corrected p-values
}</code></pre>
</div>
<div id="get-clustersmetalevels" class="section level1">
<h1 class="hasAnchor">
<a href="#get-clustersmetalevels" class="anchor"></a>GET: /Clusters/MetaLevels</h1>
<p>The display of the top-left area for signatures/proteins/meta-data depends on the selected grouping variable in the dropdown. Commands scoped to a particular grouping variable are nested under “/Clusters”.</p>
<p>This call returns a list of grouping variables and the levels/categories associated with them.</p>
<p>Response structure:</p>
<pre><code>    {
        'name of variable': [list of str - values this variable can take],
        'name of variable': [list of str - values this variable can take],
        'Cell Type': ['CD4+', 'CD8+', 'NK'],   # For example
    }</code></pre>
</div>
<div id="get-clusterssigprojmatrixnormal" class="section level1">
<h1 class="hasAnchor">
<a href="#get-clusterssigprojmatrixnormal" class="anchor"></a>GET: /Clusters/<cluster_variable>/SigProjMatrix/Normal</cluster_variable>
</h1>
<p>Returns matrices consisting of test statistic and p-value for 1 vs. all differential signature tests and for local autocorrelation.</p>
<p>First column is “Score” and values are for local autocorrelation. Statistic is the 1 - Geary’s C.</p>
<p>Other columns are for 1 vs. all differential signature tests for each value in the selected grouping variable in the dropdown. Test statistic are the AUC values from a ranksums test. P-values also from this test.</p>
<p>Response structure:</p>
<pre><code>    {
        'sig_labels': [list of str]  # Rows of matrices in zscores and pvals
        'proj_labels': [list of str] # Columns of the matrices in zscores and pvals
        'zscores': [list of list of number]  # signatures x cell group (+ 'Score') matrix of test statistics
        'pvals': [list of list of number] # signatures x cell group (+ 'Score') matrix of p-values
    }</code></pre>
</div>
<div id="get-clusterssigprojmatrixmeta" class="section level1">
<h1 class="hasAnchor">
<a href="#get-clusterssigprojmatrixmeta" class="anchor"></a>GET: /Clusters/<cluster_variable>/SigProjMatrix/Meta</cluster_variable>
</h1>
<p>Same as <code>/Clusters/&lt;cluster_variable&gt;/SigProjMatrix/Normal</code>, but for meta-data variables</p>
<p>Response structure:</p>
<pre><code>    {
        'sig_labels': [list of str]  # Rows of matrices in zscores and pvals
        'proj_labels': [list of str] # Columns of the matrices in zscores and pvals
        'zscores': [list of list of number]  # variables x cell group (+ 'Score') matrix of test statistics
        'pvals': [list of list of number] # variables x cell group (+ 'Score') matrix of p-values
    }</code></pre>
</div>
<div id="get-clustersproteinmatrix" class="section level1">
<h1 class="hasAnchor">
<a href="#get-clustersproteinmatrix" class="anchor"></a>GET: /Clusters/<cluster_variable>/ProteinMatrix</cluster_variable>
</h1>
<p>Same as <code>/Clusters/&lt;cluster_variable&gt;/SigProjMatrix/Normal</code>, but for CITE proteins</p>
<p>Response structure:</p>
<pre><code>    {
        'sig_labels': [list of str]  # Rows of matrices in zscores and pvals
        'proj_labels': [list of str] # Columns of the matrices in zscores and pvals
        'zscores': [list of list of number]  # proteins x cell group (+ 'Score') matrix of test statistics
        'pvals': [list of list of number] # proteins x cell group (+ 'Score') matrix of p-values
    }</code></pre>
</div>
<div id="get-clusterslist" class="section level1">
<h1 class="hasAnchor">
<a href="#get-clusterslist" class="anchor"></a>GET: /Clusters/list</h1>
<p>Not used anymore</p>
</div>
<div id="get-clusterscells" class="section level1">
<h1 class="hasAnchor">
<a href="#get-clusterscells" class="anchor"></a>GET: /Clusters/<cluster_variable>/Cells</cluster_variable>
</h1>
<p>Not used anymore</p>
</div>
<div id="get-expressiongeneslist" class="section level1">
<h1 class="hasAnchor">
<a href="#get-expressiongeneslist" class="anchor"></a>GET: /Expression/Genes/List</h1>
<p>Retrieves a list of genes for which expression values can be plotted. Used to populate the dropdown in the ‘Genes’ tab.</p>
<p>Response structure:</p>
<pre><code>[list of str - gene names]</code></pre>
</div>
<div id="get-expressiongene" class="section level1">
<h1 class="hasAnchor">
<a href="#get-expressiongene" class="anchor"></a>GET: /Expression/Gene/<gene_name></gene_name>
</h1>
<p>Gets the per-cell expression of the gene specified by <code>&lt;gene_name&gt;</code>. Expression values returned are on a log2 scale.</p>
<p>Response structure: Items in ‘values’ correspond to the cells in ‘cells’ in order</p>
<pre><code>    {
        'cells': [list of str - cell identifiers],
        'values': [list of number - gene expression values],
    }</code></pre>
</div>
<div id="get-cellmeta" class="section level1">
<h1 class="hasAnchor">
<a href="#get-cellmeta" class="anchor"></a>GET: /Cell/<cell_id>/Meta</cell_id>
</h1>
<p>Retrieves meta-data information for an individual cell specified by <code>&lt;cell_id&gt;</code></p>
<p>Response structure:</p>
<pre><code>    {
        'variable name': value (str or number),
        'variable name': value (str or number),
        'variable name': value (str or number),
        ...
    }</code></pre>
</div>
<div id="post-cellsmeta" class="section level1">
<h1 class="hasAnchor">
<a href="#post-cellsmeta" class="anchor"></a>POST: /Cells/Meta</h1>
<p>Retrieves meta-data information for a group of cells</p>
<p>Posted data structure</p>
<pre><code>[list of cell id]</code></pre>
<p>Response structure:</p>
<pre><code>    {
        'numeric': {
            '&lt;variable name&gt;': {
                'Min': &lt;min value&gt;,
                'Median': &lt;median value&gt;,
                'Max': &lt;max value&gt;,
            },
            '&lt;variable name&gt;': {
                'Min': &lt;min value&gt;,
                'Median': &lt;median value&gt;,
                'Max': &lt;max value&gt;,
            },
            # Repeat for additional numeric meta-data variables
        },
        'factor': {
            '&lt;variable name&gt;': {
                '&lt;factor level 1&gt;': number (proportion of cells assigned to this level,
                '&lt;factor level 2&gt;': number (proportion of cells assigned to this level,
                '&lt;factor level 3&gt;': number (proportion of cells assigned to this level,
                ...
            }
            # Repeat for additional factor meta-data variables
        }
    }</code></pre>
</div>
<div id="get-cellsselections" class="section level1">
<h1 class="hasAnchor">
<a href="#get-cellsselections" class="anchor"></a>GET: /Cells/Selections</h1>
<p>Used when loading a selection of cells.</p>
<p>Response structure:</p>
<pre><code>[list of str - names of saved selections]</code></pre>
</div>
<div id="get-cellsselections-1" class="section level1">
<h1 class="hasAnchor">
<a href="#get-cellsselections-1" class="anchor"></a>GET: /Cells/Selections/<selection_id></selection_id>
</h1>
<p>Retrieves the cells associated with the selection, <code>&lt;selection_id&gt;</code></p>
<p>Response structure:</p>
<pre><code>[list of str - cell ids in the indicated cell selection]</code></pre>
</div>
<div id="post-cellsselections" class="section level1">
<h1 class="hasAnchor">
<a href="#post-cellsselections" class="anchor"></a>POST: /Cells/Selections/<selection_id></selection_id>
</h1>
<p>Saves a cell selection with name <code>&lt;selection_id&gt;</code></p>
<pre><code>[list of str - cell ids to be assigned to this cell selection]</code></pre>
</div>
<div id="get-treeprojectionslist" class="section level1">
<h1 class="hasAnchor">
<a href="#get-treeprojectionslist" class="anchor"></a>GET: /Tree/Projections/list</h1>
<p>Get the list of different tree layouts that are available Used to populate the ‘Projection’ dropdown under the ‘Trajectories’ view</p>
<p>R version: returns the names of <a href="mailto:object@TrajectoryProjections" class="email">object@TrajectoryProjections</a></p>
<p>Response structure:</p>
<pre><code>[list of str - names of tree projections]</code></pre>
</div>
<div id="get-treeprojectionscoordinates" class="section level1">
<h1 class="hasAnchor">
<a href="#get-treeprojectionscoordinates" class="anchor"></a>GET: /Tree/Projections/<proj_name>/coordinates</proj_name>
</h1>
<p>Returns the coordinates for the tree layout <code>&lt;proj_name&gt;</code> in addition to the position of individual cells</p>
<p>Response structure:</p>
<pre><code>    [
        [
            [&lt;x val&gt;, &lt;y val&gt;, &lt;cell id&gt;],  # coordinates for individual cells
            [&lt;x val&gt;, &lt;y val&gt;, &lt;cell id&gt;],
            [&lt;x val&gt;, &lt;y val&gt;, &lt;cell id&gt;],
            ...
        ],
        [
            [&lt;x val&gt;, &lt;y val&gt;],  # coordinates for internal tree nodes
            [&lt;x val&gt;, &lt;y val&gt;], 
            [&lt;x val&gt;, &lt;y val&gt;], 
            ...
        ],
        [list of list of number]  # 0/1 adjacency matrix for internal tree nodes
    ]</code></pre>
</div>
<div id="get-treesigprojmatrixnormal" class="section level1">
<h1 class="hasAnchor">
<a href="#get-treesigprojmatrixnormal" class="anchor"></a>GET: /Tree/SigProjMatrix/Normal</h1>
<p>Returns matrices consisting of test statistic and p-value trajectory autocorrelation Format is similar to <code>/Clusters/&lt;cluster_variable&gt;/SigProjMatrix/Normal</code>, but without the 1 vs all test results.</p>
<p>‘zscores’ and ‘pvals’ matrices have just a single column</p>
<p>Response structure:</p>
<pre><code>    {
        'sig_labels': [list of str]  # Rows of matrices in zscores and pvals
        'proj_labels': ["Score"],
        'zscores': [list of list of number]  # signatures x 1 matrix of test statistics
        'pvals': [list of list of number] # signatures x 1 matrix of p-values
    }</code></pre>
</div>
<div id="get-treesigprojmatrixmeta" class="section level1">
<h1 class="hasAnchor">
<a href="#get-treesigprojmatrixmeta" class="anchor"></a>GET: /Tree/SigProjMatrix/Meta</h1>
<p>Same as <code>/Tree/SigProjMatrix/Normal</code> but for meta-data variables</p>
<p>‘zscores’ and ‘pvals’ matrices have just a single column</p>
<p>Response structure:</p>
<pre><code>    {
        'sig_labels': [list of str]  # Rows of matrices in zscores and pvals
        'proj_labels': ["Score"],
        'zscores': [list of list of number]  # variables x 1 matrix of test statistics
        'pvals': [list of list of number] # variables x 1 matrix of p-values
    }</code></pre>
</div>
<div id="get-treeproteinmatrix" class="section level1">
<h1 class="hasAnchor">
<a href="#get-treeproteinmatrix" class="anchor"></a>GET: /Tree/ProteinMatrix</h1>
<p>Same as <code>/Tree/SigProjMatrix/Normal</code> but for CITE protein expression values</p>
<p>‘zscores’ and ‘pvals’ matrices have just a single column</p>
<p>Response structure:</p>
<pre><code>    {
        'sig_labels': [list of str]  # Rows of matrices in zscores and pvals
        'proj_labels': ["Score"],
        'zscores': [list of list of number]  # proteins x 1 matrix of test statistics
        'pvals': [list of list of number] # proteins x 1 matrix of p-values
    }</code></pre>
</div>


  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>

</div>



      <footer>
      <div class="copyright">
  <p>Developed by Matt Jones, David Detomaso, Tal Ashuach, Yanay Rosen.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


