<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Functional interpretation of single cell RNA-seq latent manifolds • VISION</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><link href="extra.css" rel="stylesheet">
<meta property="og:title" content="Functional interpretation of single cell RNA-seq latent manifolds">
<meta property="og:description" content="VISION provides functional interpretation of single cell RNA-seq (scRNA-seq) latent manifolds through the use of biological signatures (which can be downloaded from online databases). VISION can operate downstream of other common analyses such as dimensionality reduction, clustering, or trajectory analysis of scRNA-seq data. VISION produces an interactive web-based output report that can be easily shared with other collaborators or the greater scientific community.">
<meta property="og:image" content="https://yoseflab.github.io/VISION/logo.svg">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">VISION</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">3.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="articles/VISION-vignette.html">Getting Started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/Signatures.html">Gene Signatures</a>
    </li>
    <li>
      <a href="articles/web_only/SignatureAutocorrelation.html">Signature Autocorrelation</a>
    </li>
    <li>
      <a href="articles/micropooling.html">Micropooling</a>
    </li>
    <li>
      <a href="articles/trajectories.html">Incorporating a Trajectory Model</a>
    </li>
    <li>
      <a href="articles/phyloVision.html">PhyloVision and Hotspot</a>
    </li>
    <li>
      <a href="articles/spatialHotspot.html">Spatial Data and Hotspot</a>
    </li>
    <li>
      <a href="articles/web_only/Seurat.html">How To: Working with Seurat</a>
    </li>
    <li>
      <a href="articles/web_only/10xData.html">How To: 10x CellRanger Outputs</a>
    </li>
  </ul>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/YosefLab/VISION/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">


<div id="functional-interpretation-for---scrna-seq-data" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#functional-interpretation-for---scrna-seq-data" class="anchor"></a>Functional Interpretation for <br> scRNA-seq Data</h1></div>
<p>VISION aids in the interpretation of single-cell RNA-seq (scRNA-seq) data by selecting for gene signatures which describe coordinated variation between cells. While the software only requires an expression matrix and a signature library (available in online databases), it is also designed to integrate into existing scRNA-seq analysis pipelines by taking advantage of precomputed dimensionality reductions, trajectory inferences or clustering results. The results of this analysis are made available through a dynamic web-app which can be shared with collaborators without requiring them to install any additional software.</p>
<ul>
<li><a href="https://www.nature.com/articles/s41467-019-12235-0">Nature Communications publication</a></li>
<li><a href="https://yoseflab.github.io/VISION/">Full Documentation</a></li>
<li><a href="https://www.biorxiv.org/content/early/2018/08/29/403055">Biorxiv Preprint</a></li>
</ul>
<div id="installing-vision" class="section level2">
<h2 class="hasAnchor">
<a href="#installing-vision" class="anchor"></a>Installing VISION</h2>
<p>We recommend installing VISION via github using devtools:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/">devtools</a></span><span class="op">)</span>
<span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span><span class="op">(</span><span class="st">"YosefLab/VISION"</span><span class="op">)</span></code></pre></div>
<p>See the DESCRIPTION file for a complete list of R dependencies. If the R dependencies are already installed, installation should finish in a few minutes.</p>
</div>
<div id="the-vision-pipeline" class="section level2">
<h2 class="hasAnchor">
<a href="#the-vision-pipeline" class="anchor"></a>The VISION Pipeline</h2>
<p>VISION generally follows the same pipeline from iteration to iteration, where minor differences can be specified via the various parameters in a VISION object. On a typical VISION run:</p>
<ul>
<li>For large datasets, or if the user so chooses, micropools are computed - grouping similar cells together to reduce the complexity of the analysis.</li>
<li>If a latent space is not specified, PCA is performed and the top 30 components are retained.</li>
<li>A KNN graph is constructed from the latent space, named the cell-cell similarity map</li>
<li>Signature scores are computed using the expression matrix</li>
<li>Signature local “consistencies” on the cell-cell similarity map are computed using the Geary-C statistic, an auto-correlation statistic.</li>
<li>An interactive web-based report is generated that can be used to explore and interpret the dataset.</li>
</ul>
</div>
<div id="how-to-run-vision" class="section level2">
<h2 class="hasAnchor">
<a href="#how-to-run-vision" class="anchor"></a>How to run VISION</h2>
<p>For general instructions on running VISION, see the <a href="https://yoseflab.github.io/VISION/articles/VISION-vignette.html">Getting Started</a> vignette.</p>
<p>More information can be found throughout the rest of the tutorials on the <a href="https://yoseflab.github.io/VISION/">Documentation</a> site.</p>
</div>
<div id="sample-output" class="section level2">
<h2 class="hasAnchor">
<a href="#sample-output" class="anchor"></a>Sample Output</h2>
<p><img src="https://github.com/YosefLab/VISION/blob/master/docs/example_report.png" alt="Screenshot of report"></p>
<ul>
<li><a href="http://s133.cs.berkeley.edu:7703/">Live output example: ~9,000 CBMC’s sequenced with the CITE-seq protocol</a></li>
<li><a href="https://github.com/YosefLab/VISION/raw/master/docs/VISION_Report_Manual.pdf">Tour of the output report user interface(PDF)</a></li>
</ul>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/YosefLab/VISION/">https://​github.com/​YosefLab/​VISION/​</a>
</li>
<li>Report a bug at <br><a href="https://github.com/YosefLab/VISION/issues">https://​github.com/​YosefLab/​VISION/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Matt Jones <br><small class="roles"> Author, maintainer </small>  </li>
<li>David Detomaso <br><small class="roles"> Author, maintainer </small>  </li>
<li>Tal Ashuach <br><small class="roles"> Author </small>  </li>
<li>Yanay Rosen <br><small class="roles"> Author </small>  </li>
<li><a href="authors.html">All authors...</a></li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://travis-ci.org/YosefLab/VISION"><img src="https://travis-ci.org/YosefLab/VISION.svg?branch=master" alt="Build status"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Matt Jones, David Detomaso, Tal Ashuach, Yanay Rosen.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
