<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to PhyloVision and Hotspot • VISION</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Introduction to PhyloVision and Hotspot">
<meta property="og:description" content="VISION">
<meta property="og:image" content="https://yoseflab.github.io/VISION/logo.svg">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">VISION</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">3.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/VISION-vignette.html">Getting Started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Signatures.html">Gene Signatures</a>
    </li>
    <li>
      <a href="../articles/web_only/SignatureAutocorrelation.html">Signature Autocorrelation</a>
    </li>
    <li>
      <a href="../articles/micropooling.html">Micropooling</a>
    </li>
    <li>
      <a href="../articles/trajectories.html">Incorporating a Trajectory Model</a>
    </li>
    <li>
      <a href="../articles/phyloVision.html">PhyloVision and Hotspot</a>
    </li>
    <li>
      <a href="../articles/spatialHotspot.html">Spatial Data and Hotspot</a>
    </li>
    <li>
      <a href="../articles/web_only/Seurat.html">How To: Working with Seurat</a>
    </li>
    <li>
      <a href="../articles/web_only/10xData.html">How To: 10x CellRanger Outputs</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/YosefLab/VISION/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="phyloVision_files/header-attrs-2.8/header-attrs.js"></script><script src="phyloVision_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introduction to PhyloVision and Hotspot</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/YosefLab/VISION/blob/master/vignettes/phyloVision.Rmd"><code>vignettes/phyloVision.Rmd</code></a></small>
      <div class="hidden name"><code>phyloVision.Rmd</code></div>

    </div>

    
    
<div id="preliminaries" class="section level1">
<h1 class="hasAnchor">
<a href="#preliminaries" class="anchor"></a>Preliminaries</h1>
<p>If you have yet to install VISION, we recommend installing the package from Github to install this package. Full source code can be found at the VISION Github repository, available <a href="http://www.github.com/YosefLab/VISION">here</a>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/">devtools</a></span><span class="op">)</span>
<span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span><span class="op">(</span><span class="st">"YosefLab/VISION"</span><span class="op">)</span></code></pre></div>
<p>Once VISION and R are installed, you may load in VISION using <code><a href="https://yoseflab.github.io/VISION">library(VISION)</a></code>.</p>
<p>Hotspot is a tool for identifying informative genes (and gene modules) in a single-cell dataset. Hotspot is installed directly from the git repository using the following command:</p>
<pre><code>pip install git+https://github.com/yoseflab/Hotspot.git</code></pre>
</div>
<div id="using-phylovision" class="section level1">
<h1 class="hasAnchor">
<a href="#using-phylovision" class="anchor"></a>Using PhyloVision</h1>
<p>Running an analysis with vision consists of three steps:</p>
<ol style="list-style-type: decimal">
<li>Creating the PhyloVision object</li>
<li>Running the <code>phyloAnalyze</code> function</li>
<li>Running Hotspot</li>
<li>Browsing results</li>
</ol>
<p>First, we need to load VISION, reticulate and ape.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">knitr</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/knitr/man/opts_chunk.html">opts_chunk</a></span><span class="op">$</span><span class="fu">set</span><span class="op">(</span>echo <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://yoseflab.github.io/VISION">VISION</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/rstudio/reticulate">reticulate</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://ape-package.ird.fr/">ape</a></span><span class="op">)</span></code></pre></div>
<div id="creating-a-phylovision-object-with-a-tree" class="section level2">
<h2 class="hasAnchor">
<a href="#creating-a-phylovision-object-with-a-tree" class="anchor"></a>Creating a PhyloVision Object with a Tree</h2>
<p>Vision objects now support dendrograms for visualization and analysis. To create the PhyloVision object, you need all of the same data, an expression matrix, and metadata and/or signatures, but you also include a tree object from the Ape package.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Read the expression and meta data</span>
<span class="va">file_path</span> <span class="op">&lt;-</span> <span class="st">"data/embryogenesis"</span> <span class="co"># Replace the path</span>

<span class="va">expr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="va">file_path</span>, <span class="st">"expr_filtered.tsv.gz"</span>, sep<span class="op">=</span><span class="st">"/"</span><span class="op">)</span>, check.names <span class="op">=</span> <span class="cn">FALSE</span>, sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span>

<span class="va">meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="va">file_path</span>, <span class="st">"meta.tsv"</span>, sep<span class="op">=</span><span class="st">"/"</span><span class="op">)</span>, check.names <span class="op">=</span> <span class="cn">FALSE</span>, sep <span class="op">=</span> <span class="st">"\t"</span>, row.names <span class="op">=</span> <span class="fl">1</span>, header<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
<span class="va">meta</span><span class="op">$</span><span class="va">Cell_Type</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">meta</span><span class="op">$</span><span class="va">Cell_Type</span><span class="op">)</span>
<span class="co"># Signature file</span>
<span class="va">sig</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"data"</span>, <span class="st">"h.all.v5.2.symbols.gmt"</span>, sep<span class="op">=</span><span class="st">"/"</span><span class="op">)</span>

<span class="co"># Read the tree</span>
<span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html">read.tree</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="va">file_path</span>, <span class="st">"embryo_tree.newick"</span>, sep<span class="op">=</span><span class="st">"/"</span><span class="op">)</span><span class="op">)</span>
<span class="co"># Collapse one mutations</span>
<span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/collapse.singles.html">collapse.singles</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span>
<span class="va">expr</span> <span class="op">&lt;-</span> <span class="va">expr</span><span class="op">[</span>, <span class="va">tree</span><span class="op">$</span><span class="va">tip.label</span><span class="op">]</span>

<span class="co"># Construct the Vision object</span>
<span class="va">vis</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PhyloVision.html">PhyloVision</a></span><span class="op">(</span>tree<span class="op">=</span><span class="va">tree</span>, data<span class="op">=</span><span class="va">expr</span>, signatures<span class="op">=</span><span class="va">sig</span>, meta<span class="op">=</span><span class="va">meta</span>, num_neighbors<span class="op">=</span><span class="fl">30</span>, projection_genes<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">expr</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><strong>Tree</strong></p>
<p>The <code>ape</code> phylogeny object. Singleton edges should be collapsed prior to use. Ensure that leaf names correspond to same names used in the expression matrix.</p>
<p><strong>Expression Data</strong></p>
<p>The provided expression data should be scaled and normalized. It is recommended to apply more advanced normalization procedures such as batch correction or removal of technical confounders.</p>
<p>The expression data should not be log-transformed prior to loading into VISION.</p>
<p><strong>Signatures</strong></p>
<p>Signatures can be provided as a list of paths to signature files (*.gmt) or Signature objects.</p>
<p>See <a href="Signatures.html">the signature vignette</a> for more information on finding or creating gene Signatures.</p>
<p><strong>Meta Data</strong></p>
<p>An R data.frame with cell-level meta-data. This could be confounding variables (e.g. percentage of mitochondrial RNA, number of genes detected) or experimental covariates (e.g. genotype, donor, batch).</p>
<p>This input is optional if Signatures are provided.</p>
<p><strong>Other Options</strong></p>
<p>Other options and inputs can be provided to customize how VISION runs.</p>
</div>
<div id="running-phylovision-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#running-phylovision-analysis" class="anchor"></a>Running PhyloVision analysis</h2>
<p>Next, we can perform the normal Vision analysis using the tree as the latent space.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">vis</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/phyloAnalyze,PhyloVision-method.html">phyloAnalyze</a></span><span class="op">(</span><span class="va">vis</span><span class="op">)</span></code></pre></div>
</div>
<div id="running-hotspot-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#running-hotspot-analysis" class="anchor"></a>Running Hotspot analysis</h2>
<p>We can also perform Hotspot module analysis. The expression data is already logged for us.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">vis</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runHotspot.html">runHotspot</a></span><span class="op">(</span><span class="va">vis</span>, model<span class="op">=</span><span class="st">"normal"</span>, tree<span class="op">=</span><span class="cn">TRUE</span>, min_gene_threshold<span class="op">=</span><span class="fl">50</span>, n_neighbors <span class="op">=</span> <span class="fl">30</span>, number_top_genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">expr</span><span class="op">)</span>, logdata<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<p>The full Hotspot API is exposed for analysis as well. For more on the Hotspot API see <a href="https://yoseflab.github.io/Hotspot/index.html">here</a></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Init Hotspot</span>
<span class="va">hs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hsInit.html">hsInit</a></span><span class="op">(</span><span class="va">vis</span>, model <span class="op">=</span> <span class="st">"normal"</span>, tree<span class="op">=</span><span class="cn">TRUE</span>, logdata<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>
<span class="co"># Init Hotspot KNN</span>
<span class="va">hs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hsCreateKnnGraph.html">hsCreateKnnGraph</a></span><span class="op">(</span><span class="va">hs</span>, <span class="va">vis</span>, n_neighbors<span class="op">=</span><span class="fl">30</span><span class="op">)</span>
<span class="co"># perform Hotspot analysis and store results in R</span>
<span class="va">hs_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hsComputeAutoCorrelations.html">hsComputeAutoCorrelations</a></span><span class="op">(</span><span class="va">hs</span>, number_top_genes<span class="op">=</span><span class="fl">12438</span><span class="op">)</span>
<span class="co"># Compute localcorr</span>
<span class="va">hs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hsComputeLocalCorrelations.html">hsComputeLocalCorrelations</a></span><span class="op">(</span><span class="va">hs</span>, <span class="va">hs_genes</span><span class="op">)</span>
<span class="co"># Calculate Hotspot Module Scores for informative genes</span>
<span class="va">hs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hsCalculateModuleScores.html">hsCalculateModuleScores</a></span><span class="op">(</span><span class="va">hs</span><span class="op">)</span>
<span class="co"># Cluster Hotspot modules and perform Vision based analysis on HS Modules and </span>
<span class="va">vis</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/analyzeHotspotObjectVision.html">analyzeHotspotObjectVision</a></span><span class="op">(</span><span class="va">vis</span>, <span class="va">hs</span>, tree<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>We can also access further Hotspot functionality using Reticulate and Python.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">hs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadHotspotObject.html">loadHotspotObject</a></span><span class="op">(</span>bytes<span class="op">=</span><span class="va">vis</span><span class="op">@</span><span class="va">Hotspot</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/rstudio/reticulate">reticulate</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/reticulate/man/use_python.html">use_python</a></span><span class="op">(</span><span class="st">'/usr/bin/python3'</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb9-2"><a href="#cb9-2"></a><span class="im">import</span> hotspot</span>
<span id="cb9-3"><a href="#cb9-3"></a>hs.plot_local_correlations()</span>
<span id="cb9-4"><a href="#cb9-4"></a>plt.show()</span></code></pre></div>
<p>Note: the heatmap can also be visualized like so:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">hs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadHotspotObject.html">loadHotspotObject</a></span><span class="op">(</span>bytes<span class="op">=</span><span class="va">vis</span><span class="op">@</span><span class="va">Hotspot</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/EmilHvitfeldt/paletteer">paletteer</a></span><span class="op">)</span>
<span class="fu"><a href="../reference/draw_hotspot_heatmap.html">draw_hotspot_heatmap</a></span><span class="op">(</span><span class="va">hs</span><span class="op">)</span></code></pre></div>
</div>
<div id="viewing-results" class="section level2">
<h2 class="hasAnchor">
<a href="#viewing-results" class="anchor"></a>Viewing Results</h2>
<p>Finally, we can launch the Vision browser.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/viewResults,Vision-method.html">viewResults</a></span><span class="op">(</span><span class="va">vis</span><span class="op">)</span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Matt Jones, David Detomaso, Tal Ashuach, Yanay Rosen.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
